cpmpy/solvers/TEMPLATE.py

```diff
@@ -397,6 +398,8 @@ def add(self, cpm_expr_orig):
             else:
                 raise NotImplementedError("TEMPLATE: constraint not (yet) supported", cpm_expr)
 
+            self.expr_dict[cpm_expr] = BoolVal(True) # constraint is now always true, no need to post it again
```

Comment:
now I'm confused...

the expr_dict is a cse_map that maps a constraint, to a decision variable representing that expression; right?

so you don't need to every overwrite it? it will just each time return that decision variable...

I might have confused you with my remark about testing transform, but the way it was was the way it should be I think... its just a map from cons to vars, so we don't need to create that equality again...

Reply:
So here I save the constraint mapping to "True" as we post the constraint to the solver.
Hence, anytime the constraint is used later-on as a subexpression to another constraint, we can safely replace it with True

Reply:
With this change I'm indeed also putting the top-level constraints into the dict, without that addition in the solver-interfaces, it is indeed a mapping from sub-expressions to auxiliary variables.

In the case of Boolean sub-expressions, it will be a Boolean variable so we introduce a double-reification; in the case of Numerical expressions, we just introduce an equality constraint.

Maybe I'm taking this CSE PR too far, will rework it to be a minimal CSE implementation and add some proper testing which will also make it more clear what types of expressions we are "caching" in the dict in what transformations.